packages/core/src/api/getBlockInfoFromPos.ts

```
@@ -36,70 +36,71 @@ export function getBlockInfo(blockContainer: Node): BlockInfoWithoutPositions {
+      return $pos.before(depth);
     }
-  }
 
-  // This gets triggered when a node selection on a block is active, i.e. when
-  // you drag and drop a block.
-  if (doc.resolve(pos).parent.type.name === "blockGroup") {
-    pos++;
+    depth--;
+    node = $pos.node(depth);
   }
 
-  const $pos = doc.resolve(pos);
-
-  const maxDepth = $pos.depth;
-  let node = $pos.node(maxDepth);
-  let depth = maxDepth;
-
-  // eslint-disable-next-line no-constant-condition
-  while (true) {
-    if (depth < 0) {
-      throw new Error(
-        "Could not find blockContainer node. This can only happen if the underlying BlockNote schema has been edited."
-      );
-    }
-
+  // If the position doesn't lie within a blockContainer node, we instead find
+  // the position of the next closest one. If the position is beyond the last
+  // blockContainer, we return the position of the last blockContainer. While
+  // running `doc.descendants` is expensive, this case should be very rarely
+  // triggered as almost every position will be within a blockContainer,
+  // according to the schema.
+  const allBlockContainerPositions: number[] = [];
+  doc.descendants((node, pos) => {
     if (node.type.name === "blockContainer") {
-      break;
+      allBlockContainerPositions.push(pos);
     }
+  });
 
-    depth -= 1;
-    node = $pos.node(depth);
-  }
+  return (
+    allBlockContainerPositions.find((position) => position >= pos) ||
+    allBlockContainerPositions[allBlockContainerPositions.length - 1]
+  );
+}
+
+/**
+ * Retrieves information regarding the nearest blockContainer node in a
+ * ProseMirror doc, relative to a position.
+ * @param doc The ProseMirror doc.
+ * @param pos An integer position.
+ * @returns A BlockInfo object for the nearest blockContainer node.
+ */
+export function getBlockInfoFromPos(doc: Node, pos: number): BlockInfo {
+  const $pos = doc.resolve(getNearestBlockContainerPos(doc, pos));
+  const node = $pos.nodeAfter!;
```

let's throw an error if node is not a blockcontainer